velox/functions/sparksql/Comparisons.h

```diff
@@ -171,4 +173,41 @@ std::shared_ptr<exec::VectorFunction> makeEqualToNullSafe(
     const std::vector<exec::VectorFunctionArg>& inputArgs,
     const core::QueryConfig& config);
 
+template <typename T>
+struct EqualToFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+  // For arbitrary nested complex types.
+  FOLLY_ALWAYS_INLINE void call(
+      bool& out,
+      const arg_type<Generic<T1>>& lhs,
+      const arg_type<Generic<T1>>& rhs) {
+    static constexpr CompareFlags kFlags =
+        CompareFlags::equality(CompareFlags::NullHandlingMode::kNullAsValue);
+
+    auto result = lhs.compare(rhs, kFlags);
+    out = (result.value() == 0);
+  }
+};
+
+template <typename T>
+struct EqualNullSafeFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+  // For arbitrary nested complex types.
+  FOLLY_ALWAYS_INLINE void callNullable(
+      bool& out,
+      const arg_type<Generic<T1>>* lhs,
+      const arg_type<Generic<T1>>* rhs) {
```

Comment:
Sorry, I have another question. With generic type supported here, will the current scalar types handling like below be replaced?

```
std::shared_ptr<exec::VectorFunction> makeEqualTo(
    const std::string& functionName,
    const std::vector<exec::VectorFunctionArg>& args,
    const core::QueryConfig& /*config*/) {
  return makeImpl<Equal>(functionName, args);
}
```

Reply:
I think no, as scalar types can't be replaced by `Generic`, for example: int, int<T1> is not a valid type.

Reply:
@zhli1142015, seems `Generic<T1>` can be used for primitive types like int64, according to my test for a simple function implementation. Could you please confirm further?

Reply:
Thanks for pointing this out, yes, you are correct, Generic<T1> can be used for primitive types.
I double confirmed primitive types are still handled by original fucntions not new ones added in this PR. I think this is because function resolution sequence.
https://facebookincubator.github.io/velox/develop/scalar-functions.html#function-resolution-order

Reply:
@zhli1142015 Thanks for checking. It seems tricky if we rely on the order of registration. Would it make sense to combine the implementations of scalar type and complex type and register once?

Reply:
Looks this is how Presto regieter eq fucntions.
https://github.com/facebookincubator/velox/blame/main/velox/functions/prestosql/registration/ComparisonFunctionsRegistration.cpp#L38-L40
If you think this is tricky, I think we may regitere these with different names.

Reply:
@mbasmanova Could you offer some advice? Thanks!

Reply:
I reviewed the registration, and it looks like we need to keep the resolution order in mind to make sure the correct signature is picked out.  Please let me know if you think differently. cc: @mbasmanova @PHILO-HE